### ISSUE TITLE
History no longer representative (2.1.0 dev 3)

### ISSUE THREAD

**Issue by HazardCreative** [RotorHazard/RotorHazard#199]
Recorded history no longer closely matches realtime data. This causes problems using history to reconstruct race laps.

Version: 2.1.0 (dev 3)
Server API: 24
Node API: 21

See below:
- Realtime calculation is catching laps not represented in history (peaks too low in history to trigger lap)
- Recalculated laps from history no longer match timing exactly (timestamps for laps 4/5/6 -> laps 1/2/3 should be within 1ms; lap 5->2 is off by 42ms)
- Based on observation, realtime data is correct and history is wrong.

![image](https://user-images.githubusercontent.com/7169527/71983077-fffff580-31f3-11ea-9188-3cc4cfc4ed09.png)
![image](https://user-images.githubusercontent.com/7169527/71983167-2de53a00-31f4-11ea-96c5-01e9b3e85f6f.png)

--

This made some races very difficult to marshal and confirm:
![image](https://user-images.githubusercontent.com/7169527/71983634-39853080-31f5-11ea-821c-6fa0c5fbab7f.png)
![image](https://user-images.githubusercontent.com/7169527/71983711-5d487680-31f5-11ea-9d4e-e76fc4d711ca.png)
![image](https://user-images.githubusercontent.com/7169527/71986896-3d1bb600-31fb-11ea-9460-239a17a3576f.png)

--

I also noticed strong signal quantization:
![image](https://user-images.githubusercontent.com/7169527/71984023-098a5d00-31f6-11ea-805d-5969b377050e.png)
![image](https://user-images.githubusercontent.com/7169527/71984125-3cccec00-31f6-11ea-9cb0-1fcc5641ff77.png)

--

These issues were not present in 2.0, so something in the newer optimizations is problematic.

Database for review and analysis: [database.zip](https://github.com/RotorHazard/RotorHazard/files/4035654/database.zip)

**Comment by pulquero** [comment]
Do we know if the unit tests pass?

**Comment by ethomas997** [comment]
We probably want to revert https://github.com/RotorHazard/RotorHazard/commit/1012eac64bc8efe0b856261a6b8bf33225d497d8

**Comment by MoonKIE666** [comment]
I agree with Michael, the other day I tried this latest version and it gave me problems to calibrate. The signal peaks did not look as clear as in previous versions. And I had to have a good time with MARSHAL to fix all the rounds before passing the results to my teammates

**Comment by ethomas997** [comment]
Did commit to revert that div-by-2 change:  https://github.com/RotorHazard/RotorHazard/commit/643ddb63047ce48f5d03f6bcf343b6e4d25bae22

Updating nodes to this version should (hopefully) fix the issue.

**Comment by pulquero** [comment]
Test has now been added to stop any future regression.

**Comment by HazardCreative** [comment]
I did some bench testing and this appears to have been resolved. Closing unless new evidence becomes available.

**Comment by HazardCreative** [comment]
This happened today. 

![image](https://user-images.githubusercontent.com/7169527/73128116-b0158280-3f98-11ea-8185-599d4ce46b5f.png)

Overall things are working much better than when I reported the issue, but is there something else in play here as well? I thought the data was perfectly represented before.

**Comment by ethomas997** [comment]
I've seen effects like that on the marshal graphs before, where the peaks look lower than they should, but I don't know if it was with particular versions.  Might be best at this point to open up a new GitHub issue and see if we can come with ways to reproduce the problem.

**Comment by HazardCreative** [comment]
I don't recall seeing this until after 2.0.2. Seems like the same issue to me?

I'll fire up RH with the RH_RECORD_NODE_0 environment var set. Maybe that will get us some data to work with.

**Comment by ethomas997** [comment]
My sense is that the original problem for this issue was the "div-by-2" being added in the node code, and the div-by-2 was removed by commit https://github.com/RotorHazard/RotorHazard/commit/1012eac64bc8efe0b856261a6b8bf33225d497d8

**Comment by HazardCreative** [comment]
I'd updated my nodes after the revert commit.

**Comment by ethomas997** [comment]
I'm just saying that I think the original cause for this issue (199) was the "div-by-2" being added in the node code; whereas what's happening now is something different.

**Comment by HazardCreative** [comment]
https://github.com/RotorHazard/RotorHazard/blob/master/src/interface/RHInterface.py#L220-L228

https://github.com/RotorHazard/RotorHazard/blob/master/src/node/commands.cpp#L136-L141

Correct me if I'm wrong: It appears the system is now collecting either a peak or a nadir for each READ_LAP_STATS polling, never both. It also appears that in order to process a peak, we must see:
- a valid peak AND an invalid nadir, OR
- a valid peak AND the start of the collected peak happened before the start of the collected nadir

The effect is that if a nadir is collected first during a polling interval, the peak is thrown out. 

Arguably the peak should have priority, but I think it's best if we continue to keep both. The reconstruction algorithm should have everything to work with that the node had.

If we want to reduce the amount of data, further node filtering should probably happen before both history collection and real-time crossing calculations occur.

**Comment by pulquero** [comment]
It should be equivalent or was intended to be equivalent to the original code, only that it takes twice as many polls to transfer the same amount of data. Maybe there is a bug or the alternative as you say is despite the frequent polling and filtering there is still sufficient noise to cause nadir/peak data to be lost. (I think there should be a bunch of 'Ignoring corrupted peak history' messages in the log in this case).

I think I am right in saying that the history collection doesnt care about peak/nadirs, it just wants a representative sample of rssi data. The node accomplishes that by picking peak/nadirs. It would be nice to replace PeakNadirHistory with RssiHistory at some point. This might help abit as we would no longer care about peaks followed by nadirs anyone - we are just collecting rssi samples. Ofcourse this does nothing for any that are lost. Quickest short term fix for that would be to try to increase the polling rate. The I2C message size has reduced a bit size over time, but the polling rate hasn't seen a corresponding increase so there is scope there. Median term fix would be to try to buffer more peak/nadirs on the node so less likely to loss any. Long term fix is to improve the filtering, but this could be alot closer than you think.

**Comment by HazardCreative** [comment]
It also looks like you start filling the buffer immediately and negate a previous peak as soon as RSSI begins rising. This may also be a problem. Within a polling window, a peak may rise, drop, and rise again. If you ignore the previous peak as soon as you start collecting a new one, you may be removing the higher peak. While it's not ideal to throw one away at all, it's very unlikely that multiple laps will occur within a polling window. The higher peak is definitely the most important, as would be the lowest nadir. That's what this check was for:

https://github.com/RotorHazard/RotorHazard/blob/2.0.2/src/node/rssi.cpp#L77

Is there a similar check in the refactor? I haven't found it but I have to admit I'm still wrapping my head around this method of saving the data.

---

Peaks are critically important to history accuracy. With a representative sample, the recalibration could reconstruct a race but there would be some flux in timing. By collecting peaks, we ensure that we have the exact data points where the crossing was strongest. The result of using those points is 1:1 agreement between realtime and recalculated race with the same values. As far as I'm aware, this level of accuracy isn't possible by doing blind sampling at any polling rate below the node looptime. On my 4-node timer my pi's sample rate is about 3Hz. Node loop time is 1KHz. That's a huge gap to make up, so I tried to build something smarter.

Nadirs are not time-critical, but their value (and which peaks they are between) is. It would be enough to send a nadir value between peaks and leave off the nadir time. For example, a data frame could be nadirValue, peakValue, peakFirst, peakLast. The nadir value here would always be the one found before the peak. (I didn't go that route in my first implementation simply because I already had a peak algorithm and inverting it was perfectly reasonable.)

**Comment by HazardCreative** [comment]
Reading your comment again, you're entirely right once the data enters python. At that point the peak and nadir items simply serve as time + value data points and are stripped of their peak/nadir designations. The PeakNadirHistory block in BaseHardwareInterface could easily be rssiHistory. The sanity checks here look to see if the timing of any data point overlaps. If they did, something was obviously wrong. Those checks will never go wrong under normal conditions, even if you throw out data points during collection. The rest of the code there really just serves to put the data points into the history store in the correct order (since either one could have occurred first in the node sample.) This could also be done through a sort function as long as it preserves the value index ordering when sorted as well.

**Comment by HazardCreative** [comment]
https://docs.google.com/spreadsheets/d/1NsLNUoAyfqwozFlXzkDsmkxwqCibokJJ6Z3C0vc2Dw8/edit?usp=sharing

Node dump. Collected history in second sheet. I've added highlights in a few places where RSSI reaches a higher level in the realtime sampling output than is ever sent as a collected history data point.

I've read through the code several times now and I think I finally understand it. Throwing out a higher peak after collecting a lower one (and the inverse for nadirs) appears to be the biggest problem this is susceptible to. We must keep the highest peak/lowest nadir we've found until it gets collected. On higher-noise nodes and systems with a greater number of nodes, this could happen pretty frequently within the node polling window.

One thing I'm realizing my original code could also have a problem with is if the second peak is the same height. Unless we buffer more data points, the best behavior would probably be to extend the peak and then discard the nadir if it was in between the two.
